CODE_GENERATION.md

```diff
@@ -0,0 +1,290 @@
+  if err != nil {
+    return managed.ExternalObservation{}, err
+  }
+  cr.SetConditions(xpv1.Available())
+  return obs, nil
+}
+```
+
+## Custom Parts for Beta Quality
+
+The earlier section is sufficient to get a controller into a working state. But
+there are some functionalities missing like late-initialization and update. These
+are not supported by ACK and even if there is support for update, it usually has
+some caveats that cannot be made generic. So, we need to implement them manually.
+
+### Late Initialization
+
+Late initialization means that if a user doesn't have a desired value for a field
+and provider chooses a default, we need to fetch that value from provider and
+assign it to the corresponding field in `spec`. You can take a look at [`RDSInstance`](https://github.com/crossplane/provider-aws/blob/087f587/pkg/clients/rds/rds.go#L368)
+example to see how it looks like but in most cases, we have much less fields so
+it's not that long.
+
+In order to override default no-op late initialization code, we'll use the same
+mechanism as others:
+```golang
+func SetupStage(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(svcapitypes.StageGroupKind)
+	opts := []option{
+		func(e *external) {
+			e.lateInitialize = lateInitialize
+			...
+		},
+	}
+	....
+func lateInitialize(*svcapitypes.StageParameters, *svcsdk.DescribeStageOutput) error {
+  return nil
+}
+```
+
+### Update
+
+#### IsUpToDate
+
+Before issuing an update call, Crossplane needs to know whether resource is up-to-date
+or not. In order to do that, we write a comparison function between observation
+and our `spec`. Here is an [example](https://github.com/crossplane/provider-aws/blob/4318043/pkg/clients/eks/nodegroup.go#L166).
+
+After writing the function, you can use it instead of the default `alwaysUpToDate`
+similar to others:
+```golang
+func SetupStage(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(svcapitypes.StageGroupKind)
+	opts := []option{
+		func(e *external) {
+			e.isUpToDate = isUpToDate
+			...
+		},
+	}
+	....
+	func isUpToDate(*svcapitypes.Stage, *svcsdk.DescribeStageOutput) bool {
+return true
+}
+```
+
+#### Update
+
+ACK has partial support for update calls. You need to inspect what's generated
+and inject custom code wherever necessary. In a lot of cases, Update calls are tricky
+and may have to be done by several calls. You can see an injected [example here](https://github.com/crossplane/provider-aws/pull/485/files#diff-8a38ae8bcdbb080fbc35feb71701c1ada99de0c36c474d42899d1ce628cf3927R277)
+with custom logic to work around an API quirk.
+
+## Testing
+
+You need to thoroughly test the controller by using example YAMLs. Make sure
+every functionality you implement works by checking with AWS console. Once all looks
+good, add the example YAML you used to `examples/<serviceid>/<crd-name>.yaml`
```

Comment:
What are our unit testing practices for ACK-generated controllers? It makes sense to skip unit tests for generated code, but should we be writing tests for the custom parts that are written by humans?

Reply:
Just added a section there. I think we need to write tests for every custom code we put in but other than that, tests for generated code should be OK to skip.